### ISSUE TITLE
Can't create IsInstantiable classes which contain reference to IOs

### ISSUE THREAD

**Issue by mwachs5** [chipsalliance/chisel3#2760]
**Type of issue**: Bug Report

**Please provide the steps to reproduce the problem:**

Scastie demonstrating the problem: https://scastie.scala-lang.org/8uH7xiDYQZqRxEokgf5Rnw
```
import chisel3._
import chisel3.experimental.hierarchy.{Definition, Instance, instantiable, public}
import chisel3.util.MixedVec

@instantiable
class LeafInstantiable(val bundle: Data){
  @public val bundle = bundle
}

@instantiable
class NestedInstantiable(val in: LeafInstantiable, val out: LeafInstantiable){
  @public val in = in
  @public val out = out
}


@instantiable
class AddOne(width: Int) extends Module {
  @public val in  = IO(Input(UInt(width.W)))
  @public val out = IO(Output(UInt(width.W)))
  out := in + 1.U

  @public val leafOut = new LeafInstantiable(out)
  @public val leafIn = new LeafInstantiable(in)
  @public val nested = new NestedInstantiable(in = leafIn, out = leafOut)
  
}

class AddTwo(width: Int) extends Module {
  val in  = IO(Input(UInt(width.W)))
  val out = IO(Output(UInt(width.W)))
  val addOneDef = Definition(new AddOne(width))
  val i0 = Instance(addOneDef)
  val i1 = Instance(addOneDef)
  i0.in := in
  i1.in := i0.out
  out   := i1.out
  
  // Doesn't work though it's equivalent to the above
  i1.leafIn.bundle := i0.leafOut.bundle
  // Also Doesn't work
  i1.nested.in.bundle := i0.nested.out.bundle
}


println(getVerilogString(new AddTwo(4)))
```

**What is the current behavior?**
```
// chisel3.internal.ChiselException:
  // Connection between sink (AddOne.mixedVecIn.0: [UInt<32>]) and
  //                    source (AddOne.mixedVecOut.0: [UInt<32>])
  // failed @: mixedVecIn.0 in AddOne cannot be written from module AddOne.
```

Seems like there may be two issues : they are not recognized as IOs on the Instance, and they might not realize the parent correctly.

**What is the expected behavior?**

I should be able to define @instantiable / IsInstantiable structures which have references to `IO`s and have those work properly e.g. connect to the IOs once I instantiate the Instance (as they do if I directly annotate them as @public).

**Please tell us about your environment:**
    - version: `3.5.4`

**Other Information**

**What is the use case for changing the behavior?**

It is convenient to mark structured data which includes IOs as @public and not have to mark the IOs public separately from related data.

**Comment by mwachs5** [comment]
3a0b8476a84f77439ce02e7164e743600b7fd707 has start of a fix

**Comment by mwachs5** [comment]
fixed by #2761
